### PR TITLE
planner: support HashJoin cost detail 

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/36962

Problem Summary:

### What is changed and how it works?

Support cost detail for hashjoin

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cost_detail.go [hunk 1]
```diff
@@ -26,23 +26,65 @@ const (
...
 	// NetworkFactorLbl indicates networkFactor
 	NetworkFactorLbl = "networkFactor"
 	// SeekFactorLbl indicates seekFactor
 	SeekFactorLbl = "seekFactor"
 	// ScanFactorLbl indicates for scanFactor
 	ScanFactorLbl = "scanFactor"
+	// SelectionFactorLbl indicates selection factor
+	SelectionFactorLbl = "selectionFactor"
+	// CPUactorLbl indicates cpu factor
+	CPUactorLbl = "cpuFactor"
```
Comment (hawkingrei) [inline planner/core/plan_cost_detail.go:45]: ```CPUFactorLBL```

**Review by hawkingrei — APPROVED** [review]

**Comment by wuhuizuo** [comment]
/gray-debug rerun because of goproxy network issue

File: util/tracing/opt_trace.go [hunk 2]
```diff
@@ -232,44 +232,48 @@ func (tracer *OptimizeTracer) RecordFinalPlan(final *PlanTrace) {
...
 // PhysicalPlanCostDetail indicates cost detail
 type PhysicalPlanCostDetail struct {
-	id     int
-	tp     string
-	params map[string]interface{}
-	desc   string
+	ID     int                    `json:"id"`
+	TP     string                 `json:"type"`
+	Params map[string]interface{} `json:"params"`
+	Desc   string                 `json:"desc"`
```
Comment (time-and-fate) [inline util/tracing/opt_trace.go:238]: Why do we export them if we already have setter methods.

File: planner/core/plan_cost_detail.go [hunk 3]
```diff
@@ -134,3 +176,177 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
...
+		AddParam(ProbeCostDescLbl, cpuCostDetail.probeCostDesc())
+	// record memory cost detail
+	detail.AddParam(MemCostDetailLbl, memoryCostDetail).
+		AddParam(MemCostDescLbl, memoryCostDetail.desc())
+	// record disk cost detail
+	detail.AddParam(DiskCostDetailLbl, diskCostDetail).
+		AddParam(DiskCostDescLbl, diskCostDetail.desc()).
+		AddParam(ProbeDiskCostDescLbl, diskCostDetail.probeDesc())
+
+	detail.SetDesc(fmt.Sprintf("%s+%s+%s+all children cost", CPUCostDetailLbl, MemCostDetailLbl, DiskCostDetailLbl))
```
Comment (time-and-fate) [inline planner/core/plan_cost_detail.go:239]: Why don't we record the cost of children operators like previous PRs?

File: planner/core/plan_cost_detail.go [hunk 4]
```diff
@@ -134,3 +176,177 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
 		RowCountLbl, RowSizeLbl, NetworkFactorLbl, NetSeekCostLbl, ScanConcurrencyLbl))
 	opt.appendPlanCostDetail(detail)
 }
+
+func setPhysicalHashJoinCostDetail(p *PhysicalHashJoin, opt *physicalOptimizeOp, spill bool,
```
Comment (time-and-fate) [inline planner/core/plan_cost_detail.go:180]: Why some factors are passed in as parameters, and some are fetched through `sessVars` inside the function?

File: util/tracing/opt_trace.go (hunk 2 above)
Comment (Yisaer) [reply util/tracing/opt_trace.go:238]: We export these params in order to transform in json later

File: planner/core/plan_cost_detail.go (hunk 3 above)
Comment (Yisaer) [reply planner/core/plan_cost_detail.go:239]: Cost Detail is used to cost operator self cost, as all operators would be recorded cost and cost detail, thus we don't need to record child cost in cost detail.

File: planner/core/plan_cost_detail.go [hunk 5]
```diff
@@ -134,3 +176,177 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
...
+	ProbeCost           *HashJoinProbeCostDetail `json:"probeCost"`
+	HashJoinConcurrency uint                     `json:"hashJoinConcurrency"`
+	Spill               bool                     `json:"spill"`
+	Cost                float64                  `json:"cost"`
+}
+
+func (h *HashJoinCPUCostDetail) desc() string {
+	var cpuCostDesc string
+	buildCostDesc := fmt.Sprintf("%s*%s", BuildRowCountLbl, CPUFactorLbl)
+	if h.Spill {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply planner/core/plan_cost_detail.go]: updated.

File: planner/core/plan_cost_detail.go (hunk 4 above)
Comment (Yisaer) [reply planner/core/plan_cost_detail.go:180]: updated.

File: planner/core/plan_cost_detail.go [hunk 6]
```diff
@@ -134,3 +176,175 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
...
+	Spill               bool                     `json:"spill"`
+	Cost                float64                  `json:"cost"`
+	UseOuterToBuild     bool                     `json:"useOuterToBuild"`
+}
+
+func (h *HashJoinCPUCostDetail) desc() string {
+	var cpuCostDesc string
+	buildCostDesc := fmt.Sprintf("%s*%s", BuildRowCountLbl, CPUFactorLbl)
+	cpuCostDesc = fmt.Sprintf("%s+%s+(%s+1)*%s)", buildCostDesc, ProbeCostDetailLbl, HashJoinConcurrencyLbl, ConcurrencyFactorLbl)
+	if h.UseOuterToBuild {
```
Comment (time-and-fate) [inline planner/core/plan_cost_detail.go:268]: You forget to set this value.

File: util/tracing/opt_trace.go (hunk 2 above)
Comment (time-and-fate) [reply util/tracing/opt_trace.go:238]: OK. Then I think we need to check the actual JSON result in the tests in the future.

File: planner/core/plan_cost_detail.go [hunk 7]
```diff
@@ -134,3 +176,175 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
...
+		AddParam(ProbeCostDescLbl, cpuCostDetail.probeCostDesc())
+	// record memory cost detail
+	detail.AddParam(MemCostDetailLbl, memoryCostDetail).
+		AddParam(MemCostDescLbl, memoryCostDetail.desc())
+	// record disk cost detail
+	detail.AddParam(DiskCostDetailLbl, diskCostDetail).
+		AddParam(DiskCostDescLbl, diskCostDetail.desc()).
+		AddParam(ProbeDiskCostDescLbl, diskCostDetail.probeDesc())
+
+	detail.SetDesc(fmt.Sprintf("%s+%s+%s+all children cost", CPUCostDetailLbl, MemCostDetailLbl, DiskCostDetailLbl))
```
Comment (time-and-fate) [inline planner/core/plan_cost_detail.go:239]: I notice that in previous PRs, each label in the `Desc` corresponds to exactly one value in the `Param` with the same label. But this PR doesn't follow this.
For example, we didn't explain the relationship between `CPUCostDetailLbl`, `CPUCostDescLbl`, and `ProbeCostDescLbl`.
Would we improve this sometime?

Comment (Yisaer) [reply planner/core/plan_cost_detail.go:240]: Currently no, we may ignore this in interval server and revise it in dashboard.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**